Effect generic bounds and functions

yoshuawuyts opened the pull request:
This explains how effect generic bounds and functions would work. I'm putting this up as a draft, since it's not yet ready - at a minimum I want to make sure the lowering section is included before we can merge this. But realistically it also needs another pass to make it flow a bit better. Thanks!

[Rendered](https://github.com/rust-lang/keyword-generics-initiative/blob/effect-generic-bounds-and-functions/explainer/effect-generic-bounds-and-functions.md)

programmerjake commented on explainer/effect-generic-bounds-and-functions.md:
```diff
@@ -0,0 +1,544 @@
+`#[maybe(async)]` annotations. This becomes more interesting once we start
+looking to not only allow traits to be generic over the `async` effect, but
+*functions* and *trait bounds* as well. The way we can do that is fairly
+*mechanical: all we have to do is add some extra `#[maybe(async)]` notations to
```
unbalanced `*`

programmerjake commented on explainer/effect-generic-bounds-and-functions.md:
```diff
@@ -0,0 +1,544 @@
+pub trait Iterator {
+    type Item;
+    #[maybe(async)]
+    fn next(&mut self) -> Option<Self::Item>;
```
what happens when you want a pinned `async Iterator`:

```rust
trait AsyncIterator {
    type Item;
    fn poll_next(self: Pin<&mut Self>) -> Poll<Option<Self::Item>>;
}
```

afaict using `async` in this style makes that difficult.

yoshuawuyts replied:
The signature for pinned iterators under this model would be:

```rust
#[maybe(async)]
pub trait PinnedIterator {
    type Item;
    #[maybe(async)]
    fn next(self: Pin<&mut Self>) -> Option<Self::Item>;
```

This not only covers the async case, but also the non-async case. Pinning however has its own set of problems. Pin projections are an essential part of the pinning system, but exist outside of the type system. So ideally we'd be able to enable functions to be polymorphic over the pinning guarantees too - enabling us to write "maybe pin" functions or similar. But that's its own topic which we shouldn't get into on this PR.

yoshuawuyts commented on explainer/effect-generic-bounds-and-functions.md:
````diff
@@ -0,0 +1,544 @@
+see RFC 0000's section on "Lowering" for more details. Now for our function
+body, the base definition looks like this:
+
+```rust
+#[maybe(async)]
+pub fn sum<I>(iter: I) -> u32
+where
+    I: #[maybe(async)] Iterator<Item = u32>
+{
+    let mut total = 0;
+    while let Some(n) = iter.next().await {
+        total += n;
+    }
+    total
+}
+```
+
+TODO: ask oli for more details about the desugaring. Ref is: https://github.com/yoshuawuyts/maybe-async-channel/blob/2fc6fa012830525482d62a8facfae5e5a5e762fe/maybe-async-std/src/lib.rs#L47-L54
````
@oli-obk this is the code I would like us to spell out the lowering for. Could you do that?

oli-obk commented on explainer/effect-generic-trait-declarations.md:
```diff
@@ -1,7 +1,7 @@
 - Start Date: (2024-01-01)
-- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
-- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+- RFC PR: [rust-lang/rfcs#0001](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0001](https://github.com/rust-lang/rust/issues/0000)
```
these links point to the rust-lang/rfcs repo, where that number is already used

clarfonthey commented on explainer/effect-generic-bounds-and-functions.md:
```diff
@@ -0,0 +1,544 @@
+#[maybe(async)] 
+pub fn copy<R, W>(reader: &mut R, writer: &mut W) -> Result<()>
+where
+    R: #[maybe(async)] Read + ?Sized,
+    W: #[maybe(async)] Write + ?Sized,
```
This isn't something to fix in the draft, but mostly just a comment: this kind of further cements the fact that the placeholder attribute syntax isn't ideal, since it's not 100% clear that the maybes are *all* linked, and that you can only have async-all or sync-all, and not a mixed async and sync.

yoshuawuyts replied:
Yeah fair point; that was also an issue brought up in #45. We probably need to find a way to better address that.

yoshuawuyts commented:
This PR has been up for a little while and has received a fair bit of feedback so far. The "lowering" section is still marked TODO, so I've added a disclaimer to it:

> _NOTE: this section is incomplete and in-progress. While we know it is possible
because we have implemented a working version of this outside of the compiler,
there are changes in the compiler happening which means this section may change.
Once those changes have landed, this section should be rewritten to match that.
Until that time please consider this section incomplete and subject to change._

With that I think we should probably merge this so it doesn't forever just sit in a PR. And then we can finish up the remaining TODO sections later on. Thanks!
